### ISSUE TITLE
A note about the program's environment

### ISSUE THREAD

**Issue by Brokolis** [oeed/Silica#1]
Hey, I noticed that you started to work on this project and I already have a design suggestion that should be applied in the beginning of the project, if applied at all.

One thing I disliked Bedrock about was that everything was being put in the global environment. This is really great when dealing with multiple files, but I had some problems where I had to restart the CC-computer to get the code to reload while developing. Also, this meant that if multiple Bedrock programs were ran at the same time in the same environment they overrode some classes and other variables which then broke programs. Another issue was security - password text boxes. You could easily find the textbox and it's value when the program is running.

My suggestion is simple: make a separate environment whenever the program loads so nothing would be put into the global environment.

And good luck on the project! That sketch Oeed put on the forums looks really great!

**Comment by viluon** [comment]
Agreed!

Offtopic: I wonder why nobody considered the option of Lua 5.2 or even 5.3 being used in the new game/new mod version...

**Comment by oeed** [comment]
Yeah, we're certainly going to make a better loading system, don't worry.

I guess we'll have to wait and see regarding 5.2/5.3.

**Comment by Brokolis** [comment]
In Lua 5.2 there are no setfenv and getfenv so that makes it harder when working with function environments. You should probably start on supporting _ENV, and maybe getfenv/setfenv for backwards compatibility with CC <= 1.7*, if you're planning to support older version of CC at all.

**Comment by oeed** [comment]
At the moment we're only making it for Lua 5.1, is the next version going to be 5.2?

**Comment by viluon** [comment]
Nobody actually knows. Have a look at the beta testing thread, dan200 talked about the switch but he didn't know environments got changed, so I suppose he didn't really think about it beforehand. (That may be because of the confusion related to minor version number change and lost of backwards compatibility, which is against semantic versioning AFAIK)

**Comment by oeed** [comment]
Well if it changes we'll change over completely.

Supporting old versions will not be happening at all with version one. Simply due to the fact that it relies on the new resolution it's completely pointless support anything older.

**Comment by viluon** [comment]
Agreed, but dan200 didn't post any additional information about the Lua version. It might be that he wants some kind of backwards compatibility with the existing programs. There could be a compatibility layer for Lua 5.1 (which is technically possible and not too hard to implement) that would simulate the CraftOS environment we're used to.

**Comment by oeed** [comment]
Well yeah. At this stage we're not really needing to interface with environments really. The loader in startup is completely temporary. But yeah, we'll need to think about that.

**Comment by viluon** [comment]
Just a thought but if you'd need setfenv and getfenv in 5.2 you could keep a table of _ENVs indexed with the functions and then just let set/getfenv to look up or change the table entry.

``` Lua
local function ASD()end
local ASDsEnv={}

Environments={[ASD]=ASDsEnv}

ASD._ENV=Environments[ASD]
```

**Comment by Brokolis** [comment]
Well, Lua 5.2 -> Lua 5.1 would be easy to do by a compatibility layer, although Lua 5.1 -> Lua 5.2 is harder. Lets just hope Dan keeps a compatibility layer for Lua 5.1.

**Comment by viluon** [comment]
_bump_ Looks like we'll have Lua 5.2 after all... http://www.computercraft.info/2015/06/27/computercraft-1-74/

**Comment by Brokolis** [comment]
Yes, but if you're not using function environments too extensively then I wrote some snippets of code for dealing with both, Lua 5.1 and 5.2:

Setting the environment of the current program/file:

``` lua
local _ENV = getCustomEnvironment() --// set the environment here \\--

if setfenv then
  setfenv(1, _ENV)
end
```

Loading a Lua chunk as a string:

``` lua
local env = getCustomEnvironment()
local func, err = (loadstring or load)("return 1", "chunk name", nil, env)

if func and setfenv then
  setfenv(func, env)
end
```

**Comment by oeed** [comment]
The new class system mainly solves this, although there might be the odd variable still in _G, I'll fix this soon.

**Comment by oeed** [comment]
I've prevented all the leaks. However, classes are still put in the global environment. The only place the global environment is actually accessible though is in the startup file (i.e. below).

``` lua
Application.static:load( "" )

QuartzApplication( ... ):run()
```

Due to the system I've made it's pretty much impossible to avoid. Besides, it's basically impossible to run any outside of classes, so it doesn't really matter.

You're welcome to reopen this, especially if it's causing issues when using Silica. For now though I don't intend to change much.
